### ISSUE TITLE
CUPS disabling opt-out

### ISSUE THREAD

**Issue by boredsquirrel** [secureblue/secureblue#33]
The current state is an OS that can't print. We should discuss why this is necessary, especially if a strict firewall is applied by default

**Comment by RoyalOughtness** [comment]
cups can be added back so long as it's disabled by default, which is not the case upstream. So additional changes will be needed to disable it in the image before it's added back

**Comment by boredsquirrel** [comment]
Collecting sources why it is actually dangerous right now would be nice. I agree best would be to have a way to enable it temporarily.

I think this sounds like the perfect job for "just".

I found this [german article](https://www.heise.de/news/Drucksystem-CUPS-Codeschmuggel-aus-dem-Netz-moeglich-9061315.html) about a big security issue which was fixed pretty quickly.

**Comment by RoyalOughtness** [comment]
> Collecting sources why it is actually dangerous right now would be nice. I agree best would be to have a way to enable it temporarily.

It increases attack surface and I suspect is unused by a majority of users. That alone is enough for it to be disabled by default. 

This project is about proactive security, not just reactive security. We want to protect against **unknown** vulnerabilities, not just quickly patch known ones.

So even using the phrase "Collecting sources why it is actually dangerous right now" demonstrates that you haven't grasped this yet. 

> I think this sounds like the perfect job for "just".

No, just is for post-install actions that can't be included in the build. This can be included in the build, so just is not relevant.

**Comment by boredsquirrel** [comment]
Task:
- [ ] Create CUPS on/off script

I am pretty close, but havent tested it yet. This is the current version:

```
#!/bin/bash

ERROR='echo -e "\033[1;31mfailed!\033[0m"'
SUCCESS='echo -e "\033[1;34mSuccess!\033[0m"'

enable_cups() {
    # Open ports for CUPS (631/tcp for printer sharing)
    echo "Opening port 631 permanently" &&\
    firewall-cmd --add-service=cups --permanent &&\
    echo "done." &&\
    echo "Reloading the firewall" &&\
    firewall-cmd --reload &&\
    echo "done." || $ERROR

    # Unmask and enable CUPS systemd service
    echo "Enabling the print service." &&\
    sudo systemctl unmask cups &&\
    sudo systemctl enable cups &&\
    sudo systemctl start cups || $ERROR
}

disable_cups() {
    # Close ports for CUPS
    echo "Closing port 631 permanently"
    firewall-cmd --remove-service=cups --permanent
    echo "Reloading the firewall"
    firewall-cmd --reload

    # Stop, disable, and mask CUPS systemd service
    echo "Disabling the print service"
    sudo systemctl stop cups
    sudo systemctl disable cups
    sudo systemctl mask cups

    # Print output, with useful explanations
    echo "CUPS disabled and Firewall port 631 closed."
    echo "You can reboot if you want to be sure."
    echo "Checking if it is still running..."

    # Check if it is inactive.
    sudo systemctl restart cups

    # we may add a "repeat function if it"
}

welcome_dialog() {

    # print in green color using ANSI escape codes
    echo -e "'\033[0;32m'------------- Secureblue hardening -------------'\033[0m'"

    # main welcome
    cat <<EOF
This script allows you to disable enable CUPS, the
printing service in Linux.

Cups can be a big security vulnerability, even
allawing remote execution. These where issues in the past,
they are fixed for now and especially in this Distro,
but there can be new vulnerabilities.

Secureblue disables CUPS by default, to protect its
users. But in a trusted environment, you are to decide
your threat model.

Press Ctrl+C to exit.
EOF
}

main() {
    welcome_dialog
    # read input for Setting
    echo "Disable or enable CUPS (Printing service): on/off"
    read -p "Turn CUPS: " action

    case "$action" in
        "on" || "enable")
            enable_cups
            echo "CUPS enabled and Firewall port 631 opened."
            # ADD EXPLANATION TO TROUBLESHOOT HERE
            ;;
        "off" || "disable")
            disable_cups
            ;;
        *)
            echo "Invalid action. Use 'on' or 'off'."
            echo
            echo "Press Ctrl+C to exit."
            ;;
    esac
}

# Run the main function
main
```

I would like to establish scripts like these, with the Welcome dialog being very verbose and empowering.

**Comment by RoyalOughtness** [comment]
> Task:
> 
> * [ ]  Create CUPS on/off script
> 
> I am pretty close, but havent tested it yet. This is the current version:
> 
> ```
> #!/bin/bash
> 
> ERROR='echo -e "\033[1;31mfailed!\033[0m"'
> SUCCESS='echo -e "\033[1;34mSuccess!\033[0m"'
> 
> enable_cups() {
>     # Open ports for CUPS (631/tcp for printer sharing)
>     echo "Opening port 631 permanently" &&\
>     firewall-cmd --add-service=cups --permanent &&\
>     echo "done." &&\
>     echo "Reloading the firewall" &&\
>     firewall-cmd --reload &&\
>     echo "done." || $ERROR
> 
>     # Unmask and enable CUPS systemd service
>     echo "Enabling the print service." &&\
>     sudo systemctl unmask cups &&\
>     sudo systemctl enable cups &&\
>     sudo systemctl start cups || $ERROR
> }
> 
> disable_cups() {
>     # Close ports for CUPS
>     echo "Closing port 631 permanently"
>     firewall-cmd --remove-service=cups --permanent
>     echo "Reloading the firewall"
>     firewall-cmd --reload
> 
>     # Stop, disable, and mask CUPS systemd service
>     echo "Disabling the print service"
>     sudo systemctl stop cups
>     sudo systemctl disable cups
>     sudo systemctl mask cups
> 
>     # Print output, with useful explanations
>     echo "CUPS disabled and Firewall port 631 closed."
>     echo "You can reboot if you want to be sure."
>     echo "Checking if it is still running..."
> 
>     # Check if it is inactive.
>     sudo systemctl restart cups
> 
>     # we may add a "repeat function if it"
> }
> 
> welcome_dialog() {
> 
>     # print in green color using ANSI escape codes
>     echo -e "'\033[0;32m'------------- Secureblue hardening -------------'\033[0m'"
> 
>     # main welcome
>     cat <<EOF
> This script allows you to disable enable CUPS, the
> printing service in Linux.
> 
> Cups can be a big security vulnerability, even
> allawing remote execution. These where issues in the past,
> they are fixed for now and especially in this Distro,
> but there can be new vulnerabilities.
> 
> Secureblue disables CUPS by default, to protect its
> users. But in a trusted environment, you are to decide
> your threat model.
> 
> Press Ctrl+C to exit.
> EOF
> }
> 
> main() {
>     welcome_dialog
>     # read input for Setting
>     echo "Disable or enable CUPS (Printing service): on/off"
>     read -p "Turn CUPS: " action
> 
>     case "$action" in
>         "on" || "enable")
>             enable_cups
>             echo "CUPS enabled and Firewall port 631 opened."
>             # ADD EXPLANATION TO TROUBLESHOOT HERE
>             ;;
>         "off" || "disable")
>             disable_cups
>             ;;
>         *)
>             echo "Invalid action. Use 'on' or 'off'."
>             echo
>             echo "Press Ctrl+C to exit."
>             ;;
>     esac
> }
> 
> # Run the main function
> main
> ```
> 
> I would like to establish scripts like these, with the Welcome dialog being very verbose and empowering.

Looks like good progress! If this can be integrated into yafti like you mentioned, then we can allow the user to select their preference at install time

**Comment by boredsquirrel** [comment]
also it seems that CUPS might simply have way too high privileges?

It just needs to print I mean... accept a job, send data, thats it.

It feels like this should really be investigated further, after first disabling it by default and then allowing the users to opt-in and also have the manual option to do so.

Might be out of scope? But a system that can't print, because the implementation sucks, it pretty bad

maybe isolating it in a rootless Podman container could work, but only after these work

**Comment by wandapeter** [comment]
couldn't we simply package the `enable_cups()` as a script in `/usr/libexec` and then have a yafti consent screen which explains the risks and calls this script? 

I feel like we have all the required components to create the opt-out.

**Comment by RoyalOughtness** [comment]
> couldn't we simply package the `enable_cups()` as a script in `/usr/libexec` and then have a yafti consent screen which explains the risks and calls this script?
> 
> I feel like we have all the required components to create the opt-out.

Yes, all we need is:

1. config to disable cups by default
2. a script to enable it
3. yafti config to allow the user to enable it through the GUI

This should be very quick and straightforward, it just needs to get done.

**Comment by trytomakeyouprivate** [comment]
I am curious about how to apply systemd changes like these onto the image by default. Is this possible, does this need to be done on first startup? This would allow maybe some more optimizations, like force enabling automatic updates which seem to not always work on ublue (?)

**Comment by RoyalOughtness** [comment]
https://github.com/secureblue/secureblue/commit/195d2e90ff5d79a07b8681419e972f5d2836ef1f
